File: practice/sorting/go_code/bubbleSort.go
package main

import "fmt"

func main() {
	unsorted := []int{10, 6, 2, 1, 5, 8, 3, 4, 7, 9}
	sorted := bubbleSort(unsorted)
	fmt.Println(sorted)
	// sorted = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
}

func bubbleSort(arr []int) []int {
	swapped := true
	for swapped {
		swapped = false
		for i := 1; i < len(arr); i++ {
			if arr[i-1] > arr[i] {
				arr[i], arr[i-1] = arr[i-1], arr[i]
				swapped = true
			}
		}
	}
	return arr
}

File: practice/sorting/python_code/bubble_sort.py
"""
Bubble Sort
The Bubble sort algorithm compares each pair of elements in an array and swaps them if they are out of order until the entire array is sorted. For each element in the list, the algorithm compares every pair of elements.

By convention, empty arrays and singleton arrays (arrays consisting of only one element) are always sorted. This is a key point for the base case of many sorting algorithms.

Source: https://brilliant.org/wiki/bubble-sort/
"""

from typing import List


def bubble_sort(arr: List[int]) -> List[int]:
    """Bubble Sort"""
    if len(arr) == 1 or arr is None:
        return arr
    index = len(arr) - 1
    while index >= 0:
        for j in range(index):
            if arr[j] > arr[j + 1]:
                arr[j + 1], arr[j] = arr[j], arr[j + 1]
        index -= 1
        print(arr, index)

    return arr


def bubble_sort_recursive(arr: List[int], n: int) -> List[int]:
    if n <= 1:
        return arr

    for i in range(n - 1):
        if arr[i] > arr[i + 1]:
            arr[i], arr[i + 1] = arr[i + 1], arr[i]

    bubble_sort_recursive(arr, n - 1)
    return arr


if __name__ == "__main__":
    arr = [12, 4, 8, 2, 15, 13, 1]
    # print(bubble_sort(arr))
    print(bubble_sort_recursive(arr, len(arr)))
